token/src/token.rs

```diff
@@ -0,0 +1,164 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use zkvm::{
+    Commitment, Data, Entry, Input, Instruction, Output, Predicate, Program, Prover, Signature, Tx,
+    TxHeader, TxID, TxLog, VMError, Value, VerificationKey, Verifier,
+};
+
+/// Represents a ZkVM Token with unique flavor and embedded
+/// metadata protected by a user-supplied Predicate.
+#[derive(Clone, Debug)]
+pub struct Token {
+    issuance_predicate: Predicate,
+    metadata: Vec<u8>,
+}
+
+impl Token {
+    /// Constructs a new Token.
+    pub fn new(pred: Predicate, metadata: Vec<u8>) -> Self {
+        Token {
+            issuance_predicate: pred,
+            metadata: metadata,
+        }
+    }
+
+    /// Returns the Token's flavor.
+    pub fn flavor(&self) -> Scalar {
+        Value::issue_flavor(
+            &self.issuance_predicate,
+            Data::Opaque(self.metadata.clone()),
+        )
+    }
+
+    /// Returns program that issues specified quantity of Token.
```

Comment:
This sounds a bit misleading: sounds like it has no side-effects. It does return a program reference only for chaining. But it _modifies_ the program with issuance logic.

Maybe smth like:

```suggestion
    /// Adds instructions to a program that issue a given quantity of this token.
```